src/network/networkd-netdev-wireguard.c

```diff
@@ -0,0 +1,625 @@
+
+    systemd is distributed in the hope that it will be useful, but
+    WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sys/ioctl.h>
+#include <net/if.h>
+
+#include "alloc-util.h"
+#include "parse-util.h"
+#include "in-addr-util.h"
+#include "fd-util.h"
+#include "strv.h"
+#include "hexdecoct.h"
+#include "string-util.h"
+#include "networkd-netdev-wireguard.h"
+#include "networkd-link.h"
+#include "networkd.h"
+
+#define WG_SET_DEVICE (SIOCDEVPRIVATE + 1)
+#define MAX_CONCURRENT_QUERIES 2048
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *new_peer(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        if (!peer)
+                return NULL;
+        peer->fields.replace_ipmasks = true;
+        peer->fields.persistent_keepalive_interval = (uint16_t)-1;
+
+        LIST_PREPEND(peers, w->peers, peer);
+        w->allocation_size += sizeof(struct wgpeer);
+        w->last_peer_section = section;
+        w->dev.num_peers++;
+
+        return peer;
+}
+
+static int update_wireguard_config(NetDev *netdev) {
+        _cleanup_close_ int fd = -2;
+        _cleanup_free_ void *data = NULL;
+        void *pos = NULL;
+        struct ifreq ifreq = {0};
+        size_t len;
+        Wireguard *w;
+        WireguardPeer *peer;
+        WireguardIpmask *mask;
+
+        assert(netdev);
+
+        w = WIREGUARD(netdev);
+
+        assert(w);
+
+        fd = socket(AF_INET, SOCK_DGRAM, 0);
+        if (fd < 0)
+                return log_netdev_error_errno(netdev, -errno, "Failed to open AF_INET to configure wireguard: %m");
+
+        len = strlen(netdev->ifname);
+        assert(len <= IFNAMSIZ);
+
+        memcpy(&ifreq.ifr_name, netdev->ifname, len + 1);
+        data = calloc(sizeof(char), w->allocation_size);
+        if (!data)
+                return log_netdev_error_errno(netdev, -errno, "Cannot allocate memory to configure wireguard: %m");
+
+        pos = data;
+        *((struct wgdevice*)pos) = w->dev;
+        pos = (struct wgdevice*)pos + 1;
+        LIST_FOREACH(peers, peer, w->peers) {
+                *((struct wgpeer*)pos) = peer->fields;
+                pos = (struct wgpeer*)pos + 1;
+
+                LIST_FOREACH(ipmasks, mask, peer->ipmasks) {
+                        *((struct wgipmask*)pos) = mask->fields;
+                        pos = (struct wgipmask*)pos + 1;
+                }
+        }
+        assert((size_t)((char*)pos - (char*)data) == w->allocation_size);
+        ifreq.ifr_data = data;
```

Comment:
hmm, uh oh... the whole loop raises all eyebrows i have due to alignment concerns... I think it would be nicer to use memcpy() when putting this together...

(and I'd probably increase the ptr, which "pos = (uint8_t*) pos + siezof(struct wgdevice)" and suchlike, I think that'd be more readable...